controllers/util/common/common.go

```diff
@@ -0,0 +1,208 @@
+	for _, crb := range clusterRoleBindings {
+		_, err := ch.getClusterRoleBinding(crb)
+		if err != nil && errors.IsNotFound(err) {
+			logger.Info("Creating a new ClusterRoleBinding", "Name", crb.GetName())
+			err = ch.Create(context.TODO(), crb)
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			logger.Error(err, "Failed to get ClusterRole", "Name", crb.GetName())
+			return err
+		} else {
+			// Resource already exists - don't requeue
+			//ch.Log.Info("Skip reconcile: ClusterRoleBinding already exists", "Name", crb.GetName())
+		}
+	}
+	return nil
+}
+
+func (ch *ControllerHelper) getClusterRoleBinding(crb *rbacv1.ClusterRoleBinding) (*rbacv1.ClusterRoleBinding, error) {
+	found := &rbacv1.ClusterRoleBinding{}
+	err := ch.Get(context.TODO(), types.NamespacedName{
+		Name:      crb.Name,
+		Namespace: crb.Namespace,
+	}, found)
+	return found, err
+}
+
+func (ch *ControllerHelper) DeleteClusterRoles(clusterRoles []*rbacv1.ClusterRole) error {
+	logger := ch.Log.WithName("DeleteClusterRoles")
+	for _, cr := range clusterRoles {
+		found, err := ch.getClusterRole(cr)
+		if err != nil && errors.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			logger.Error(err, "failed to get ClusterRole", "Name", cr.GetName())
+			return err
+		} else {
+			logger.Info("deleting ClusterRole", "Name", cr.GetName())
+			if err := ch.Delete(context.TODO(), found); err != nil {
+				logger.Error(err, "failed to delete ClusterRole", "Name", cr.GetName())
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (ch *ControllerHelper) ReconcileClusterRole(clusterRoles []*rbacv1.ClusterRole) error {
+	logger := ch.Log.WithValues("Resource Type", "ClusterRole")
+	for _, cr := range clusterRoles {
+		_, err := ch.getClusterRole(cr)
+		if err != nil && errors.IsNotFound(err) {
+			logger.Info("Creating a new ClusterRole", "Name", cr.GetName())
+			err = ch.Create(context.TODO(), cr)
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			logger.Error(err, "Failed to get ClusterRole", "Name", cr.GetName())
+			return err
+		} else {
+			err = ch.Update(context.TODO(), cr)
+			if err != nil {
+				logger.Error(err, "Failed to update ClusterRole", "Name", cr.GetName())
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (ch *ControllerHelper) getClusterRole(cr *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+	found := &rbacv1.ClusterRole{}
+	err := ch.Get(context.TODO(), types.NamespacedName{
+		Name:      cr.GetName(),
+		Namespace: cr.GetNamespace(),
+	}, found)
+	return found, err
+}
+
+func (ch *ControllerHelper) HasFinalizer(instance crutils.Instance) (bool, error) {
+	accessor, finalizerName, err := ch.getAccessorAndFinalizerName(instance)
+	if err != nil {
+		return false, err
+	}
+
+	return util.Contains(accessor.GetFinalizers(), finalizerName), nil
+}
+
+func (ch *ControllerHelper) AddFinalizerIfNotPresent(instance crutils.Instance,
+	unwrappedInstance client.Object) error {
```

Comment:
why do we need `unwrappedInstance`? can't we instance.Unwrap()?
do we ever not call it with (instance, instance.Unwrap())?

Reply:
because I couldn't replace the [return type](https://github.com/IBM/ibm-block-csi-operator/blob/bd57e05c5479cd0fdde51092ba799072c2552f59/controllers/internal/crutils/ibmblockcsi.go#L43) to [`crutils.Instance`](https://github.com/IBM/ibm-block-csi-operator/blob/bd57e05c5479cd0fdde51092ba799072c2552f59/controllers/internal/crutils/common.go#L25), so I couldn't add an Unwrap function to the `crutils.Instance` interface and we can't use the Unwrap function in `AddFinalizerIfNotPresent` function